File: 377-combination-sum-iv/377-combination-sum-iv.py
class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
	
		# Sort the nums to break the loop when total - num becomes negative.
        nums.sort()

		# Allocates the dp table
        dp = [0] * (target + 1)

        # The case total becomes 0 is only 1 (nothing add)
        dp[0] = 1

		# Try the all cases the total from 1 to target
        for currtotal in range(1, target + 1):
		
			# Try all possible numbers.
            for number in nums:
			
                if currtotal - number >= 0:
					# Uses the previous information
                    dp[currtotal] += dp[currtotal - number]
                else:
                    break
        return dp[target]

File: 377-combination-sum-iv/377-combination-sum-iv.java
class Solution {
    public int combinationSum4(int[] nums, int target) {
        int[] dp=new int[target+1];
        Arrays.fill(dp,-1);
        return ways(nums,target, dp);
    }
    int ways(int[] nums, int target, int dp[]){
        
        if (target==0) return 1;
        
        if (target<0) return 0;
        
        if(dp[target]!=-1) return dp[target];
        
        int res=0;
        for (int i=0;i<nums.length;i++){
            res += ways(nums, target-nums[i], dp);
        }
        
        dp[target]=res;
        return dp[target];
    }
}